lib/di/interceptor/app_interceptor.dart

```diff
@@ -28,48 +31,59 @@ class AppInterceptor extends Interceptor {
       options.headers
           .putIfAbsent(_authorizationHeader, () => '$tokenType $accessToken');
     }
-    return super.onRequest(options, handler);
+    super.onRequest(options, handler);
   }
 
   @override
   void onError(DioError err, ErrorInterceptorHandler handler) {
-    handler.next(err);
-    // TODO: Integrate refresh-token https://github.com/nimblehq/ic-flutter-taher-toby/issues/20
-    // final statusCode = err.response?.statusCode;
-    // if ((statusCode == HttpStatus.forbidden ||
-    //         statusCode == HttpStatus.unauthorized) &&
-    //     _requireAuthentication) {
-    //   _doRefreshToken(err, handler);
-    // } else {
-    // }
+    final statusCode = err.response?.statusCode;
+    if ((statusCode == HttpStatus.forbidden ||
+            statusCode == HttpStatus.unauthorized) &&
+        _requireAuthentication) {
+      _doRefreshToken(err, handler);
+    } else {
+      handler.next(err);
+    }
   }
 
-  // ignore: unused_element
   Future<void> _doRefreshToken(
-    DioError err,
+    DioError error,
     ErrorInterceptorHandler handler,
   ) async {
     try {
-      // if (result is Success) {
-      // err.requestOptions.headers[_headerAuthorization] = newToken;
-      // Create request with new access token
-      final options = Options(
-          method: err.requestOptions.method,
-          headers: err.requestOptions.headers);
-      final newRequest = await _dio.request(
-          "${err.requestOptions.baseUrl}${err.requestOptions.path}",
+      final RefreshTokenUseCase refreshTokenUseCase =
+          getIt.get<RefreshTokenUseCase>();
+      final StoreAuthTokenUseCase storeAuthTokenUseCase =
+          getIt.get<StoreAuthTokenUseCase>();
```

Comment:
We should get and pass the instance from `dio_provider` to this `AppInterceptor` to avoid object re-creating 🙏 We should get it in each `AppInterceptor.onError` call as it will slow down the performance

Reply:
I tried to pass it from Dio but, I got error for  `RefreshTokenUseCase` and `StoreAuthTokenUseCase`
<img width="452" alt="Screenshot 2023-04-24 at 2 47 38 PM" src="https://user-images.githubusercontent.com/13636465/233932320-eac4a180-4b82-46e5-a441-54dde04348cf.png">

**Error is**:

[VERBOSE-2:dart_vm_initializer.cc(41)] Unhandled Exception: 'package:get_it/get_it_impl.dart': Failed assertion: line 372 pos 7: 'instanceFactory != null': Object/factory with  type RefreshTokenUseCase is not registered inside GetIt.
get_it_impl.dart:372
(Did you accidentally do GetIt sl=GetIt.instance(); instead of GetIt sl=GetIt.instance;
Did you forget to register it?)

Reply:
@Shayokh144 I could be wrong, could it be we must clean & build again?

>  fvm flutter clean && fvm flutter pub get && fvm flutter packages pub run build_runner build --delete-conflicting-outputs

Reply:
@Tuubz I tried this but same error

Reply:
I think you can find the same approach on Android, for example https://github.com/nimblehq/worpt-android/blob/1163064b693fd2daed02bac802e9c93a2d347027/app/src/main/java/com/worpt/wallet/di/modules/AppModule.kt#L21-L24

Reply:
@luongvo I tried by making the usecases singleton but get the same error.

Reply:
@Shayokh144 Please push up a draft commit when you're available 🙇

Reply:
@luongvo @Tuubz I pushed a [commit here](https://github.com/nimblehq/ic-flutter-taher-toby/pull/75/commits/0be15c8c1d30c224eacf24cc4faae82e42c6ab8c)

Reply:
@luongvo Any idea? I know you're busy with Gulf, please have a look when you're available 🙏 

I tried myself as well, I wasn't able to find any way to make it work 🥶

Reply:
@Shayokh144 Let's proceed with your initial solution here, we can adapt later if needed (when Lucas is back) 🙏